File: pptx-service/renderers/persona.py
from builder import add_bg, add_textbox, add_rect, add_line, contrast_color
from theme import ThemeColors

OPEN_QUOTE = "“"
CLOSE_QUOTE = "”"


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    name = data.get("name", "")
    role = data.get("role", "")
    attributes = data.get("attributes", []) or []
    quote = data.get("quote", "")
    needs = data.get("needs", []) or []
    pains = data.get("pains", []) or []

    initials = name[0].upper() if name else "?"
    avatar_text = contrast_color(t.accent)
    add_rect(slide, 0.5, 0.5, 2.0, 2.0, fill_hex=t.accent)
    add_textbox(slide, 0.5, 0.9, 2.0, 1.2, initials,
                t.font_display, 80, avatar_text, bold=True, align="center")

    add_textbox(slide, 2.8, 0.6, 9.5, 0.9, name,
                t.font_display, 48, t.text, bold=True)
    add_textbox(slide, 2.8, 1.5, 9.5, 0.5, role,
                t.font_body, 24, t.muted)

    add_line(slide, 0.5, 2.7, 12.8, 2.7, t.border, 0.5)

    y = 2.9
    for attr in attributes[:4]:
        add_textbox(slide, 0.5, y, 2.5, 0.45,
                    attr.get("label", "") + ":", t.font_body, 17, t.muted, bold=True)
        add_textbox(slide, 3.0, y, 9.8, 0.45,
                    attr.get("value", ""), t.font_body, 17, t.text)
        y += 0.5

    if needs or pains:
        add_line(slide, 6.67, 4.8, 6.67, 7.3, t.border, 0.5)

        add_textbox(slide, 0.5, 4.8, 5.8, 0.5, "NEEDS",
                    t.font_body, 14, t.muted, bold=True)
        ny = 5.3
        for n_item in needs[:3]:
            add_textbox(slide, 0.5, ny, 5.8, 0.5,
                        f"+ {n_item}", t.font_body, 18, t.text, wrap=True)
            ny += 0.5

        add_textbox(slide, 7.0, 4.8, 5.8, 0.5, "PAIN POINTS",
                    t.font_body, 14, t.muted, bold=True)
        py2 = 5.3
        for p_item in pains[:3]:
            add_textbox(slide, 7.0, py2, 5.8, 0.5,
                        f"- {p_item}", t.font_body, 18, t.text, wrap=True)
            py2 += 0.5

    if quote:
        add_line(slide, 0.5, 6.8, 12.8, 6.8, t.border, 0.3)
        add_textbox(slide, 0.5, 6.9, 12.0, 0.55,
                    OPEN_QUOTE + quote + CLOSE_QUOTE,
                    t.font_body, 18, t.muted, italic=True)

File: lib/pptx/slides/persona.ts
import type { PersonaSlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addLine, addTextBox, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: PersonaSlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);

  const avatarSize = px(120);
  addRect(slide, t.paddingIn, py(80), avatarSize, avatarSize, '#' + t.accent);
  const initials = data.name.split(' ').slice(0, 2).map(w => w[0] ?? '').join('').toUpperCase();
  addTextBox(slide, initials, {
    x: t.paddingIn, y: py(80), w: avatarSize, h: avatarSize,
    fontSize: Math.round(t.heroSize * 0.4), color: 'FFFFFF',
    fontFace: t.fontDisplay, bold: true, align: 'center', valign: 'middle',
  });

  const infoX = t.paddingIn + avatarSize + px(30);
  const infoW = SLIDE_W * 0.45 - avatarSize - px(30);

  addTextBox(slide, data.name, { x: infoX, y: py(90), w: infoW, fontSize: t.sectionSize, color: '#' + t.text, fontFace: t.fontDisplay, bold: true });
  addTextBox(slide, data.role, { x: infoX, y: py(170), w: infoW, fontSize: t.bodySize, color: '#' + t.muted, fontFace: t.fontBody });

  addLine(slide, t.paddingIn, py(250), SLIDE_W - t.paddingIn, py(250), '#' + t.rule, 1);

  const attrs = data.attributes ?? [];
  attrs.slice(0, 6).forEach((attr, i) => {
    const col = i % 2;
    const row = Math.floor(i / 2);
    const attrX = t.paddingIn + col * (SLIDE_W * 0.45);
    const attrY = py(280) + row * py(100);
    addTextBox(slide, attr.label.toUpperCase(), { x: attrX, y: attrY, w: px(300), fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody });
    addTextBox(slide, attr.value, { x: attrX, y: attrY + py(36), w: px(300), fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody });
  });

  const rightX = SLIDE_W * 0.55;
  const rightW = SLIDE_W - rightX - t.paddingIn;
  let ry = py(260);

  if (data.needs && data.needs.length > 0) {
    addTextBox(slide, 'NEEDS', { x: rightX, y: ry, w: rightW, fontSize: t.captionSize, color: '#' + t.accent, fontFace: t.fontBody });
    ry += py(40);
    data.needs.forEach(need => {
      addTextBox(slide, `✓ ${need}`, { x: rightX, y: ry, w: rightW, fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, wrap: true });
      ry += py(70);
    });
  }

  if (data.pains && data.pains.length > 0) {
    ry += py(20);
    addTextBox(slide, 'PAINS', { x: rightX, y: ry, w: rightW, fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody });
    ry += py(40);
    data.pains.forEach(pain => {
      addTextBox(slide, `✗ ${pain}`, { x: rightX, y: ry, w: rightW, fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, wrap: true });
      ry += py(70);
    });
  }

  if (data.quote) {
    addLine(slide, t.paddingIn, py(760), t.paddingIn, py(880), '#' + t.accent, 4);
    addTextBox(slide, `"${data.quote}"`, {
      x: t.paddingIn + px(20), y: py(770), w: SLIDE_W * 0.5,
      fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, italic: true, wrap: true,
    });
  }

  addPageNumber(slide, n, total, t);
}
